frontend/catalyst/pennylane_extensions.py

```diff
@@ -465,6 +467,107 @@ def _check(x, hint):
     return jprim.vjp_p.bind(*params, *cotangents, jaxpr=jaxpr, fn=fn, grad_params=grad_params)
 
 
+class Adjoint(Operation):
+    """A minimal implementation of PennyLane operation, designed with a sole purpose of being
+    placed on the quantum tape"""
+
+    num_wires = AnyWires
+
+    def __init__(self, body_jaxpr, consts_jaxpr, cargs_jaxpr):
+        self.body_jaxpr = body_jaxpr
+        self.consts_jaxpr = list(consts_jaxpr)
+        self.cargs_jaxpr = list(cargs_jaxpr)
+        super().__init__(wires=Wires(Adjoint.num_wires))
+
+
+def adjoint(f: Union[Callable, Operator]) -> Union[Callable, Operator]:
+    """A :func:`~.qjit` compatible adjoint transformer for PennyLane/Catalyst.
+
+    Returns a quantum function or operator that applies the adjoint of the
+    provided function or operator.
+
+    .. warning::
+
+        This function does not support performing the adjoint
+        of quantum functions that contain Catalyst control flow
+        or mid-circuit measurements.
+
+    Args:
+        f (Callable or Operator): A PennyLane operation or a Python function
+                                  containing PennyLane quantum operations.
+
+    Returns:
+        If an Operator is provided, returns an Operator that is the adjoint. If
+        a function is provided, returns a function with the same call signature
+        that returns the Adjoint of the provided function.
+
+    Raises:
+        ValueError: invalid parameter values
+
+    **Example**
+
+    .. code-block:: python
+
+        @qjit
+        @qml.qnode(qml.device("lightning.qubit", wires=1))
+        def workflow(theta, wires):
+            catalyst.adjoint(qml.RZ)(theta, wires=wires)
+            catalyst.adjoint(qml.RZ(theta, wires=wires))
+            def func():
+                qml.RX(theta, wires=wires)
+                qml.RY(theta, wires=wires)
+            catalyst.adjoint(func)()
+            return qml.probs()
+
+    >>> workflow(pnp.pi/2, wires=0)
+    array([0.5, 0.5])
+    """
+
+    def _trace_quantum_tape(*args, _callee: Callable):
+        (qargs, cargs, ckwargs) = args
+        assert len(qargs) == 1
+        with JaxTape(do_queue=False) as tape:
+            with tape.quantum_tape:
+                out = _callee(*cargs, **ckwargs)
```

Comment:
Well, I just used the pattern I found elsewhere, might need to check carefully what happens here.

Reply:
Just a note here that instead of `do_queue=False` we should use:

> with qml.QueuingManager.stop_recording():

as do_queue=False is being deprecated.

Reply:
Ah yes, I think Erick is right. That is also what the other classes in this file use.